crates/extensions/tedge_flows/src/registry.rs

```diff
@@ -0,0 +1,237 @@
+use crate::config::ConfigError;
+use crate::config::FlowConfig;
+use crate::flow::Flow;
+use crate::js_runtime::JsRuntime;
+use async_trait::async_trait;
+use camino::Utf8Path;
+use camino::Utf8PathBuf;
+use std::collections::HashMap;
+use tracing::error;
+use tracing::info;
+use tracing::warn;
+
+#[async_trait]
+pub trait FlowRegistry {
+    type Flow: Send + AsRef<Flow> + AsMut<Flow>;
+
+    fn compile(flow: Flow) -> Result<Self::Flow, ConfigError>;
+
+    fn store(&self) -> &FlowStore<Self::Flow>;
+    fn store_mut(&mut self) -> &mut FlowStore<Self::Flow>;
+
+    fn deadlines(&self) -> impl Iterator<Item = tokio::time::Instant> + '_;
+}
+
+pub struct BaseFlowRegistry {
+    flows: FlowStore<Flow>,
+}
+
+impl BaseFlowRegistry {
+    pub fn new(config_dir: impl AsRef<Utf8Path>) -> Self {
+        BaseFlowRegistry {
+            flows: FlowStore::new(config_dir),
+        }
+    }
+}
+
+#[async_trait]
+impl FlowRegistry for BaseFlowRegistry {
+    type Flow = Flow;
+
+    fn compile(flow: Flow) -> Result<Flow, ConfigError> {
+        Ok(flow)
+    }
+
+    fn store(&self) -> &FlowStore<Self::Flow> {
+        &self.flows
+    }
+
+    fn store_mut(&mut self) -> &mut FlowStore<Self::Flow> {
+        &mut self.flows
+    }
+
+    fn deadlines(&self) -> impl Iterator<Item = tokio::time::Instant> + '_ {
+        self.flows()
+            .flat_map(|flow| &flow.steps)
+            .filter_map(|step| step.script.next_execution)
+    }
+}
+
+#[async_trait]
+pub trait FlowRegistryExt: FlowRegistry {
+    fn config_dir(&self) -> Utf8PathBuf;
+
+    fn flow(&self, name: &str) -> Option<&Self::Flow>;
+    fn flow_mut(&mut self, name: &str) -> Option<&mut Self::Flow>;
+
+    fn flows(&self) -> impl Iterator<Item = &Self::Flow>;
+    fn flows_mut(&mut self) -> impl Iterator<Item = &mut Self::Flow>;
```

Comment:
Optional suggestion: These functions could be moved into `FlowRegistry` itself as default implementations, right? As they only rely on the `store()` and `store_mut()` functions? That'll make the separation between `FlowRegistry` and `FlowRegistryExt` clearer as well, where the former deals with all the core access primitives and the latter deals with the dependency on `JsRuntime`.

Reply:
> These functions could be moved into `FlowRegistry` itself as default implementations, right? As they only rely on the `store()` and `store_mut()` functions?

Technically, yes. However, it makes no sense to have another implementation. Hence grouped into an `Ext` trait.

> That'll make the separation between `FlowRegistry` and `FlowRegistryExt` clearer as well, where the former deals with all the core access primitives and the latter deals with the dependency on `JsRuntime`.

This is not the intent of that separation.  `FlowRegistryExt` groups the methods for which no other implementation makes sense.